File: src/stm/history.py
"""SQLite 操作歷史:記錄每筆對收藏的 add/delete,支援查詢與標記復原。

一個使用者動作 = 一個 batch_id(一鍵刪 N 首算一批);復原以 batch 為單位反向。
純資料層,不碰 Spotify;反向操作由 server 持 client 執行。
"""

from __future__ import annotations

import sqlite3
import uuid
from datetime import datetime, timezone

_SCHEMA = """
CREATE TABLE IF NOT EXISTS operations (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    batch_id TEXT NOT NULL,
    ts TEXT NOT NULL,
    action TEXT NOT NULL,
    track_id TEXT NOT NULL,
    name TEXT,
    artist TEXT,
    undone INTEGER NOT NULL DEFAULT 0
)
"""


class History:
    def __init__(self, path: str = ":memory:"):
        self._conn = sqlite3.connect(path, check_same_thread=False)
        self._conn.row_factory = sqlite3.Row
        self._conn.execute(_SCHEMA)
        self._conn.commit()

    def record(self, action: str, tracks: list[dict], batch_id: str | None = None) -> str:
        """記錄一批同類動作;tracks 為 [{id, name, artist}]。回傳 batch_id。"""
        batch_id = batch_id or uuid.uuid4().hex
        ts = datetime.now(timezone.utc).isoformat()
        self._conn.executemany(
            "INSERT INTO operations(batch_id, ts, action, track_id, name, artist)"
            " VALUES (?, ?, ?, ?, ?, ?)",
            [
                (batch_id, ts, action, t["id"], t.get("name", ""), t.get("artist", ""))
                for t in tracks
            ],
        )
        self._conn.commit()
        return batch_id

    def list_batches(self) -> list[dict]:
        """每個 batch 一列摘要(新到舊):batch_id, action, ts, n, undone。"""
        rows = self._conn.execute(
            "SELECT batch_id, action, MIN(ts) AS ts, COUNT(*) AS n, MAX(undone) AS undone"
            " FROM operations GROUP BY batch_id ORDER BY MIN(id) DESC"
        ).fetchall()
        return [dict(r) for r in rows]

    def batch_tracks(self, batch_id: str) -> list[dict]:
        rows = self._conn.execute(
            "SELECT action, track_id, name, artist FROM operations WHERE batch_id = ?"
            " ORDER BY id",
            (batch_id,),
        ).fetchall()
        return [dict(r) for r in rows]

    def mark_undone(self, batch_id: str) -> None:
        self._conn.execute(
            "UPDATE operations SET undone = 1 WHERE batch_id = ?", (batch_id,)
        )
        self._conn.commit()

File: apps/backend/src/adapters/db/history.ts
import Database from "better-sqlite3";
import type { HistoryBatch } from "@stm/shared";
import type { HistoryStore } from "../../ports/history-store";

// A durable op-log so every library mutation can be undone. Each user action is one
// "batch" (a delete or an add of N track ids). Undo reverses the batch: a delete is
// undone by re-adding the ids, an add by removing them. The op-log only records what
// happened — the caller performs the actual Spotify reversal and reports success.

interface BatchRow {
  batch_id: string;
  action: "add" | "delete";
  ts: string;
  track_ids: string;
  undone: number;
}

const toBatch = (r: BatchRow): HistoryBatch => ({
  batchId: r.batch_id,
  action: r.action,
  ts: r.ts,
  count: (JSON.parse(r.track_ids) as string[]).length,
  undone: r.undone === 1,
});

export class History implements HistoryStore {
  private db: Database.Database;

  constructor(path = ":memory:") {
    this.db = new Database(path);
    this.db.pragma("journal_mode = WAL");
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS batches (
        batch_id  TEXT PRIMARY KEY,
        action    TEXT NOT NULL CHECK (action IN ('add','delete')),
        ts        TEXT NOT NULL,
        track_ids TEXT NOT NULL,
        undone    INTEGER NOT NULL DEFAULT 0
      );
    `);
  }

  /** Record a completed mutation. `batchId`/`ts` are supplied by the caller (Node's clock). */
  record(action: "add" | "delete", trackIds: string[], batchId: string, ts: string): void {
    this.db
      .prepare("INSERT INTO batches (batch_id, action, ts, track_ids, undone) VALUES (?, ?, ?, ?, 0)")
      .run(batchId, action, ts, JSON.stringify(trackIds));
  }

  /** All batches, newest first. */
  list(): HistoryBatch[] {
    const rows = this.db
      .prepare("SELECT batch_id, action, ts, track_ids, undone FROM batches ORDER BY ts DESC")
      .all() as BatchRow[];
    return rows.map(toBatch);
  }

  /**
   * Mark a batch undone and return what to reverse. Returns null if the batch is
   * unknown or already undone (so the caller makes no Spotify call). The reversal
   * itself (add↔delete) is performed by the caller.
   */
  beginUndo(batchId: string): { action: "add" | "delete"; trackIds: string[] } | null {
    const row = this.db
      .prepare("SELECT action, track_ids, undone FROM batches WHERE batch_id = ?")
      .get(batchId) as Pick<BatchRow, "action" | "track_ids" | "undone"> | undefined;
    if (!row || row.undone === 1) return null;
    this.db.prepare("UPDATE batches SET undone = 1 WHERE batch_id = ?").run(batchId);
    return { action: row.action, trackIds: JSON.parse(row.track_ids) as string[] };
  }

  /** Revert the undone flag — used to roll back if the Spotify reversal fails. */
  cancelUndo(batchId: string): void {
    this.db.prepare("UPDATE batches SET undone = 0 WHERE batch_id = ?").run(batchId);
  }

  close(): void {
    this.db.close();
  }
}
